libs/core/type_support/include/hpx/type_support/uninitialized_relocate.hpp

```diff
@@ -0,0 +1,162 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/type_support/is_relocatable.hpp>
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+#include <hpx/type_support/relocate_at.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+#include <memory>
+#endif
+
+namespace hpx {
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+    using std::uninitialized_relocate;
+#else
+
+    namespace detail {
+
+        template <typename InIter, typename FwdIter>
+        struct choose_uninitialized_relocate_helper
+        {
+            //     using in_type = typename std::remove_reference_t<
+            //         decltype(*std::declval<InIter>())>;
+            //     using out_type = typename std::remove_reference_t<
+            //         decltype(*std::declval<FwdIter>())>;
+
+            using in_type = typename std::iterator_traits<InIter>::value_type;
+            using out_type = typename std::iterator_traits<FwdIter>::value_type;
+
+            constexpr static bool valid_relocation =
+                hpx::is_relocatable_from_v<out_type, in_type>;
+
+            constexpr static bool is_buffer_memcpyable =
+                hpx::is_trivially_relocatable_v<in_type> &&
+                //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ The important check
+                std::is_same_v<std::decay_t<in_type>, std::decay_t<out_type>> &&
+                //  can only relocate between same types
+                !std::is_volatile_v<in_type> && !std::is_volatile_v<out_type> &&
+                //  volatile types are not memcpyable
+                std::is_pointer_v<InIter> && std::is_pointer_v<FwdIter>;
+            //  ^^ the best we can do to check for contiguous iterators
```

Comment:
We do have `hpx::traits::is_contiguous_iterator[_v]`, not sure if that's usable in this context.